internal/graph/lookupsubjects.go

```diff
@@ -495,114 +910,389 @@ func (lsu *lookupSubjectsUnion) CompletedChildOperations() error {
+	for _, foundSubjects := range subjects {
+		for _, foundSubject := range foundSubjects.FoundSubjects {
+			// NOTE: wildcard is always returned, because it is needed by all branches, at all times.
+			if foundSubject.SubjectId == tuple.PublicWildcard || (afterSubjectID == "" || foundSubject.SubjectId > afterSubjectID) {
+				filteredSubjectIDs.Add(foundSubject.SubjectId)
 			}
 		}
+	}
 
-		if index == 0 {
-			foundSubjects = results
-		} else {
-			foundSubjects.SubtractAll(results)
-			if foundSubjects.IsEmpty() {
-				return nil
-			}
+	sortedSubjectIDs := filteredSubjectIDs.AsSlice()
+	sort.Strings(sortedSubjectIDs)
+
+	subjectIDsToPublish := make([]string, 0, len(sortedSubjectIDs))
+	lastSubjectIDToPublishWithoutWildcard := ""
+
+	done := func() {}
+	for _, subjectID := range sortedSubjectIDs {
+		// Wildcards are always published, regardless of the limit.
+		if subjectID == tuple.PublicWildcard {
+			subjectIDsToPublish = append(subjectIDsToPublish, subjectID)
+			continue
 		}
+
+		ok := ci.limits.prepareForPublishing()
+		if !ok {
+			break
+		}
+
+		subjectIDsToPublish = append(subjectIDsToPublish, subjectID)
+		lastSubjectIDToPublishWithoutWildcard = subjectID
 	}
 
-	return lse.parentStream.Publish(&v1.DispatchLookupSubjectsResponse{
-		FoundSubjectsByResourceId: foundSubjects.AsMap(),
+	if len(subjectIDsToPublish) == 0 {
+		return nil, done, nil
+	}
+
+	// Determine the subject ID for the cursor. If there are any concrete subject IDs, then the last
+	// one is used. Otherwise, the wildcard itself is published as a specialized cursor to indicate that
+	// all concrete subjects have been consumed.
+	cursorSubjectID := "*"
+	if len(lastSubjectIDToPublishWithoutWildcard) > 0 {
+		cursorSubjectID = lastSubjectIDToPublishWithoutWildcard
+	}
+
+	updatedCI, err := ci.withOutgoingSection(cursorSubjectID)
+	if err != nil {
+		return nil, done, err
+	}
+
+	return &v1.DispatchLookupSubjectsResponse{
+		FoundSubjectsByResourceId: filterSubjectsMap(subjects, subjectIDsToPublish),
 		Metadata:                  metadata,
-	})
+		AfterResponseCursor:       updatedCI.responsePartialCursor(),
+	}, done, nil
+}
+
+// publishSubjects publishes the given subjects to the stream, after appying filtering and limiting.
+func publishSubjects(stream dispatch.LookupSubjectsStream, ci cursorInformation, subjects map[string]*v1.FoundSubjects) error {
+	response, done, err := createFilteredAndLimitedResponse(ci, subjects, emptyMetadata)
+	defer done()
+	if err != nil {
+		return err
+	}
+
+	if response == nil {
+		return nil
+	}
+
+	return stream.Publish(response)
+}
+
+// filterSubjectsMap filters the subjects found in the subjects map to only those allowed, returning an updated map.
+func filterSubjectsMap(subjects map[string]*v1.FoundSubjects, allowedSubjectIds []string) map[string]*v1.FoundSubjects {
+	updated := make(map[string]*v1.FoundSubjects, len(subjects))
+	allowed := mapz.NewSet[string](allowedSubjectIds...)
```

Comment:
Its passed through a set but this does additional filtering. We could likely combine it though in the method above, since it is only used there. Shall I try to do so?

Reply:
I think so